### PR TITLE
Query: Remove ReLinq dependency in Funcletizer

### LINKED ISSUE #13524: Client evaluation in query with Contains on an IEnumerable subquery
This is the query pattern that [caused StackOverflow.com to go offline yesterday](https://twitter.com/Nick_Craver/status/1047929955206795267):

image

There are two separate aspects we need to follow up on:

1. This is a predicate that we could do a better job of translating to SQL. We will use this issue to track that pattern specifically. 

2. The customer was porting their code from LINQ to SQL to EF Core and didn't expect client evaluation to occur for a predicate against a table that contained millions of rows, and did not think of [switching client-eval off](https://docs.microsoft.com/en-us/ef/core/querying/client-eval#optional-behavior-throw-an-exception-for-client-evaluation). Assuming the query was exercised in testing or staging, that could have helped detect the issue before rolling up to production. Based on this and other previous feedback, we are discussing moving the slider for client eval in future versions. Whatever shape this takes, it is likely to be a breaking change, so we will track it as part of the overall 3.0 plan for query in #12795.

## Workaround
If you are hitting client evaluation for this query, you can follow the same pattern @JoshSchreuder described below: split the sub-query into its own variable:

```csharp
var postIds = grp.Select(g => g.PostId); 
var answers = db.Posts.Where(postIds.Contains(p.Id)).Select(...).ToList();
```

### LINKED ISSUE #13549: Query: Funcletization Improvements
Find the biggest evaluatable subtree to parametrize.

### PR DESCRIPTION
**Pull request by smitpatel**
Fixes #13524
Fixes #13549

Part of #12048

### CODE REVIEW

**Comment by smitpatel** [comment]
Notes:
Disabled documentation error (since a lot of code would be changing in query area with new temporary methods)

- `EvaluatableExpressionFilter` still exists from Relinq since we use it for query parsing, hence renamed all the conflicting class names to have prefix of ReLinq.

**Comment by smitpatel** [comment]
If providers are trying to block funcletization of any expression in their codebase then they need to update to use new filter.

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs [hunk 1]
```diff
@@ -1199,6 +1199,49 @@ public virtual async Task Contains_with_local_list_inline_closure_mix(bool isAsy
                     }.Contains(c.CustomerID)), entryCount: 1);
         }
 
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Contains_with_local_non_primitive_list_inline_closure_mix(bool isAsync)
+        {
+            var id = "ALFKI";
+
+            await AssertQuery<Customer>(
```
Comment (maumar) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: consider negative case when one of the arguments is correlated with customer

**Comment by smitpatel** [comment]
@nfdevil - See https://github.com/aspnet/AspNetCore/blob/master/docs/daily-builds.md
The channel you want to pull from is master.

**Pull request merged**